### PR TITLE
feat: building dbconnections signup route

### PR DESCRIPTION
**Pull request by danielgent**
Reverse engineer `auth0.js` using PRs on login2 e.g. https://github.com/sesamyab/login2/pull/580

### CODE REVIEW

File: src/routes/tsoa/dbconnections.ts [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+  email: string;
+  email_verified: boolean;
+  app_metadata: {};
+  user_metadata: {};
+}
+
+@Route("dbconnections")
+@Tags("dbconnections")
+
+// note plural dbconnectionS - not the same as {clientId}/dbconnection
```
Comment (markusahlstrand) [inline src/routes/tsoa/dbconnections.ts:47]: Nice!

File: src/routes/tsoa/dbconnections.ts [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    if (!client) {
+      throw new HTTPException(400, { message: "Client not found" });
+    }
+
+    const existingUser = await getPrimaryUserByEmailAndProvider({
+      userAdapter: env.data.users,
+      tenant_id: client.tenant_id,
+      email,
+      // we are only allowing this on this route... I'm not sure how it could be different!
+      provider: "auth2",
```
Comment (markusahlstrand) [inline src/routes/tsoa/dbconnections.ts:74]: Think it's correct that this is hardcoded

File: src/routes/tsoa/dbconnections.ts [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    });
+
+    if (existingUser) {
+      // we can copy what auth0 returns here
+      throw new HTTPException(400, { message: "User already exists" });
```
Comment (markusahlstrand) [inline src/routes/tsoa/dbconnections.ts:79]: They're not returning a 409?

File: src/routes/tsoa/dbconnections.ts [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+    }
+
+    const primaryUser = await getPrimaryUserByEmail({
+      userAdapter: env.data.users,
+      tenant_id: client.tenant_id,
+      email,
+    });
+
+    const newUser = await env.data.users.create(client.tenant_id, {
+      id: `email|${userIdGenerate()}`,
```
Comment (markusahlstrand) [inline src/routes/tsoa/dbconnections.ts:89]: Shouldn't this be `auth2|` ?

File: src/routes/tsoa/dbconnections.ts [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+      // we are only allowing this on this route... I'm not sure how it could be different!
+      provider: "auth2",
+    });
+
+    if (existingUser) {
+      // we can copy what auth0 returns here
+      throw new HTTPException(400, { message: "User already exists" });
+    }
+
+    const primaryUser = await getPrimaryUserByEmail({
```
Comment (markusahlstrand) [inline src/routes/tsoa/dbconnections.ts:82]: At some point I would like to separate the linking to a post-create-hook.. Now we're duplicating this logic for each connection.

File: src/routes/tsoa/dbconnections.ts [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+    // TODO - I think we handle account linking LAST
+    // investigate on auth0
+    // i. sign in with a new code use dan+newcodeuser@sesamy.com
+    // ii. then register... see what happens
+
+    return {
+      _id: newUser.id,
+      email: newUser.email,
+      // this is the crux of the issue!  8-)   What to do here?
+      email_verified: false,
```
Comment (markusahlstrand) [inline src/routes/tsoa/dbconnections.ts:111]: I think.. we should create the user as non-verified and to not merge it. Then we trigger the merge when the account is verified and (and block the login until the verify it..)

**Review by markusahlstrand — APPROVED** [review]
Looks good. The only objection is that I think the provider (auth2) should be before the pipe in the id.

File: src/routes/tsoa/dbconnections.ts (hunk 3 above)
Comment (danielgent) [reply src/routes/tsoa/dbconnections.ts:79]: nope!

image

Ah, so they don't actually leak info on the user existing! Which is why it's a 400 :smile: 

I'll change the error message to this now

Comment (danielgent) [reply src/routes/tsoa/dbconnections.ts:79]:
```suggestion
      // Auth0 doesn't inform that the user already exists
      throw new HTTPException(400, { message: "Invalid sign up" });
```

Comment (markusahlstrand) [reply src/routes/tsoa/dbconnections.ts]: Makes sense :)

File: src/routes/tsoa/dbconnections.ts (hunk 5 above)
Comment (danielgent) [reply src/routes/tsoa/dbconnections.ts:83]: Of course, but done properly, so first we write the functionality and test it :smile: 

I have a feeling it's not all quite the same, but that's the joy of TDD, as opposed to navel gazing over correct code when externally it's probably all on fire :fire:

File: src/routes/tsoa/dbconnections.ts (hunk 4 above)
Comment (danielgent) [reply src/routes/tsoa/dbconnections.ts:90]: YES! Nice spot :smile:

File: src/routes/tsoa/dbconnections.ts (hunk 6 above)
Comment (danielgent) [reply src/routes/tsoa/dbconnections.ts:112]: ahhhh, of course. I was thinking that so I don't know why I did this :joy: 

:facepalm: 

My next steps are seeing what Auth0 does with email-password accounts :upside_down_face: 

I'll remove any fetching of any primary user, which is correct :+1:

**Comment by danielgent** [comment]
Cool comments addressed!

**Pull request merged**